src/query/storages/stage/src/read/row_based/formats/csv/separator.rs

```diff
@@ -0,0 +1,313 @@
+            ReadRecordResult::OutputFull => Err(self.error_output_full()),
+            ReadRecordResult::OutputEndsFull => Err(self.error_output_ends_full()),
+            ReadRecordResult::Record => {
+                let output = {
+                    if self.projection.is_some() {
+                        // select $1, $2, $3 ..  from csv, not check num of fields here
+                        ReadRecordOutput::Record {
+                            num_fields: n_end,
+                            bytes: n_out,
+                        }
+                    } else {
+                        // copy
+                        if !self.error_on_column_count_mismatch {
+                            // cut or patch to num_fields
+
+                            if self.n_end < self.num_fields {
+                                // support we expect 4 fields but got row with only 2 columns : "1,2\n"
+                                // here we pretend we read "1,2,,\n"
+                                debug_assert!(self.n_end > 0);
+                                let end = self.field_ends[n_end - 1];
+                                for i in n_end..self.num_fields {
+                                    self.field_ends[i] = end;
+                                }
+                            }
+                            ReadRecordOutput::Record {
+                                num_fields: self.num_fields,
+                                bytes: n_out,
+                            }
+                        } else {
+                            // check num of fields strictly
+
+                            if let Err(e) = self.check_num_field() {
+                                self.load_ctx.error_handler.on_error(
+                                    e,
+                                    None,
+                                    file_status,
+                                    &self.pos.path,
+                                    self.pos.rows,
+                                )?;
+                                ReadRecordOutput::RecordSkipped
+                            } else {
+                                ReadRecordOutput::Record {
+                                    num_fields: self.num_fields,
+                                    bytes: n_out,
+                                }
+                            }
+                        }
+                    }
+                };
+                self.pos.offset += n_in;
+                self.pos.rows += 1;
+                self.n_end = 0;
+                Ok((output, n_in))
+            }
+            ReadRecordResult::End => {
+                if !input.is_empty() {
+                    Err(ErrorCode::BadBytes("unexpected eof"))
+                } else {
+                    Ok((ReadRecordOutput::PartialRecord { bytes: n_out }, n_in))
+                }
+            }
+        }
+    }
+
+    fn separate(&mut self, batch: BytesBatch) -> Result<(Vec<RowBatch>, FileStatus)> {
```

Comment:
The logic of those codes is not very well understood, perhaps adding some comments would be better

Reply:
added some comments here and for the pipeline